### PR TITLE
SI-6626 make @throws tags create links to exceptions

### PR DESCRIPTION
**Pull request by gourlaysama**
In scaladoc, this turns exceptions in `@throws` tags into links
(when it can find the target exception), instead of just showing
the name.

Review by @VladUreche :)

### CODE REVIEW

File: src/scaladoc/scala/tools/nsc/doc/base/CommentFactoryBase.scala [hunk 1]
```diff
@@ -345,12 +345,30 @@ trait CommentFactoryBase { this: MemberLookupBase =>
           Map.empty[String, Body] ++ pairs
         }
 
+        def linkedExceptions: Map[String, Body] = {
+          val m = allSymsOneTag(SimpleTagKey("throws"))
+
+          m.map { case (a,b) =>
```
Comment (Kwestor) [inline src/scaladoc/scala/tools/nsc/doc/base/CommentFactoryBase.scala]: What is `a` and `b`?

File: src/scaladoc/scala/tools/partest/ScaladocModelTest.scala [hunk 2]
```diff
@@ -182,14 +182,16 @@ abstract class ScaladocModelTest extends DirectTest {
       }
     }
 
-    def countLinks(c: Comment, p: EntityLink => Boolean) = {
+    def countLinks(c: Comment, p: EntityLink => Boolean): Int = countLinks(c.body, p)
+
+    def countLinks(c: Body, p: EntityLink => Boolean): Int = {
```
Comment (Kwestor) [inline src/scaladoc/scala/tools/partest/ScaladocModelTest.scala]: To maintain convention it should be `b: Body`. I prefer `body: Body`.

File: src/scaladoc/scala/tools/nsc/doc/base/CommentFactoryBase.scala (hunk 1 above)
Comment (gourlaysama) [inline src/scaladoc/scala/tools/nsc/doc/base/CommentFactoryBase.scala]: Right, those are pretty unclear; fixed.

File: src/scaladoc/scala/tools/partest/ScaladocModelTest.scala (hunk 2 above)
Comment (gourlaysama) [inline src/scaladoc/scala/tools/partest/ScaladocModelTest.scala]: Hmm, copy/paste FTW. I went for `body`.

**Comment by VladUreche** [comment]
@gourlaysama, seems the new method prevents the presentation compiler from being compiled:

``` scala
fail - presentation/doc  [compilation failed]% scalac doc/doc.scala
doc.scala:42: error: object creation impossible, since method linkName in trait MemberLookupBase of type (link: scala.tools.nsc.doc.base.LinkTo)String is not defined
    new Global(settings, compilerReporter) with MemberLookupBase with CommentFactoryBase with doc.ScaladocGlobalTrait {
        ^
```

(see https://scala-webapps.epfl.ch/jenkins/job/pr-scala-test/7400/console)

Can you please address this?

**Comment by gourlaysama** [comment]
@VladUreche sorry about that. I should have asked for review after jenkins was finished.

I just realized this will break scala-ide too, for the same reason as above: it needs to implement that method to do the right thing...

**Comment by VladUreche** [comment]
LGTM! Now that the build passes, I can merge. Thanks @gourlaysama!

**Comment by lrytz** [comment]
Hi, question about this change: it seems that we get a warning now whenever writing `@throws E` and `E` is on the classpath, not in the project being compiled. I guess that's not intended?

```
lucmac:sandbox luc$ cat Test.scala 
class A {
  /**
   * @throws Exk bla
   * @throws java.lang.Exception
   */
  def foo = 1
}
class Exk extends Exception

lucmac:sandbox luc$ ~/scala/scala-2.11.6/bin/scaladoc Test.scala 
model contains 3 documentable templates
Test.scala:2: warning: Could not find any member to link for "java.lang.Exception".
  /**
  ^
one warning found
```

see https://issues.scala-lang.org/browse/SI-9311

**Comment by VladUreche** [comment]
@lrytz: Yes, it's expected: https://github.com/scala/scala/pull/4089/files#diff-cf5321b0df0af2c25cec233dbb01cc20R352

Ifneedbe we can silence the warning.

**Comment by lrytz** [comment]
I think we should skip the warning for exception classes that are out of the project, it's not so nice to have warnings that you cannot get rid of.

**Comment by VladUreche** [comment]
@gourlaysama, can you take care of this? Or should I do it?

**Comment by gourlaysama** [comment]
@VladUreche Yep I'll take that one. Warnings that aren't actionable are useless...

**Pull request merged**